File: heap-vs-stack/stack_overflow.cpp
#include <iostream>

void blowStack(int depth) {
    std::cout << "Recursion depth: " << depth << std::endl;
    blowStack(depth + 1);
}

void largeStackAllocation() {
    int largeArray[1000000];
    largeArray[0] = 42;
    std::cout << "Large stack allocation succeeded: " << largeArray[0] << std::endl;
}

int main() {
    std::cout << "=== Stack Overflow Example ===" << std::endl;
    std::cout << "Warning: Uncomment blowStack(0) or largeStackAllocation() to crash!" << std::endl;
    return 0;
}

File: heap-vs-stack/stack_overflow.rs
fn blow_stack(depth: u64) {
    println!("Recursion depth: {}", depth);
    blow_stack(depth + 1);
}

fn large_stack_allocation() {
    let large_array: [i32; 1_000_000] = [0; 1_000_000];
    println!("Large stack allocation succeeded: {}", large_array[0]);
}

fn deep_recursion(depth: u64) {
    let _local_var = depth;
    if depth % 1000 == 0 {
        println!("Recursion depth: {}", depth);
    }
    deep_recursion(depth + 1);
}

fn main() {
    println!("=== Stack Overflow Example ===");
    println!("Warning: This will crash!");
    println!("Uncomment blow_stack(0), large_stack_allocation(), or deep_recursion(0) to crash");
}
